src/named_tuple.cr

```diff
@@ -35,7 +35,22 @@ struct NamedTuple
   # {}             # syntax error
   # ```
   def self.new(**options : **T)
-    options
+    {% if @type.name(generic_args: false) == "NamedTuple" %}
+      # deduced type vars
+      options
+    {% elsif @type.name(generic_args: false) == "NamedTuple()" %}
+      # special case: empty named tuple
+      options
+    {% else %}
+      # explicitly provided type vars
+      {% begin %}
+        {
+          {% for key in T %}
+            {{ key.stringify }}: options[{{ key.symbolize }}].as(typeof(0.unsafe_as(self)[{{ key.symbolize }}])),
```

Comment:
Out of curiosity, what's the `0.unsafe_as(self)` about?

Reply:
Obtains a value of type `Tuple(*T)` for type deduction purposes, without having to go through any of the constructors (in C++ this would be called `std::declval`). Using `typeof` here instead of doing `{{ T[key] }}` also means we could get private types to work, e.g. most iterators returned by `Iterator`.

Reply:
why not use self instead?

Reply:
Didn't know that `Tuple.[]` exists. Let's see if it works...

Reply:
Oh, it's because you want the instance type... nevermind.

Reply:
You could also do: x = uninitialized self; x[...]

Reply:
I rmember seeing `uninitialized` in the runtime version of `Class.|`. Is `uninitialized` guaranteed to generate no code when used this way?

Both `0.unsafe_as(...)` and `x = uninitialized ...; typeof(x)` look like we're dealing with low-level code when we really aren't doing anything unsafe, so a dedicated expression for this would be nice (not in this PR, of course). But if I use the former inside a `typeof`, at least I'm not under the impression that the Crystal compiler could possibly generate some code for an uninitialized value, not even outside release mode.

Reply:
You can put the entire expression into `typeof` so no variable is created in actual code: `typeof(begin x = uninitialized self; x[{{ key.symbolize }}] end)`.

Reply:
I think I also prefer the `uninitialized` way. It's more explicit than `unsafe_as`.